### PR TITLE
[SYCL-MLIR] Generate LLVM functions in `gpu.module`

### PR DESCRIPTION
**Pull request by whitneywhtsang**
`Basic/group_local_id.cpp` passes after this PR.

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 1]
```diff
@@ -927,6 +927,10 @@ ValueCategory MLIRScanner::VisitCallExpr(clang::CallExpr *Expr) {
       }
     }
 
+  FunctionContext FuncContext =
+      isa<mlir::gpu::GPUModuleOp>(Function->getParentOp())
+          ? FunctionContext::SYCLDevice
+          : FunctionContext::Host;
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGCall.cc:933]: We have `mlirclang::getFuncContext` for this

File: polygeist/tools/cgeist/Lib/CGCall.cc [hunk 2]
```diff
@@ -1192,7 +1206,10 @@ MLIRScanner::emitGPUCallExpr(clang::CallExpr *Expr) {
 
         if (Arg.getType().isa<LLVM::LLVMPointerType>()) {
           const clang::FunctionDecl *Callee = EmitCallee(Expr->getCallee());
-          LLVM::LLVMFuncOp StrcmpF = Glob.getOrCreateLLVMFunction(Callee);
+          LLVM::LLVMFuncOp StrcmpF = Glob.getOrCreateLLVMFunction(
+              Callee, isa<mlir::gpu::GPUModuleOp>(Function->getParentOp())
+                          ? FunctionContext::SYCLDevice
+                          : FunctionContext::Host);
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/CGCall.cc:1212]: Same as above

File: polygeist/tools/cgeist/Lib/clang-mlir.cc [hunk 3]
```diff
@@ -1618,7 +1619,13 @@ MLIRASTConsumer::getOrCreateLLVMFunction(const clang::FunctionDecl *FD) {
                                                 /*isVarArg=*/FD->isVariadic());
   // Insert the function into the body of the parent module.
   OpBuilder Builder(Module->getContext());
-  Builder.setInsertionPointToStart(Module->getBody());
+  if (FuncContext == FunctionContext::SYCLDevice)
+    Builder.setInsertionPointToStart(
+        mlirclang::getDeviceModule(*Module).getBody());
+  else {
+    assert(FuncContext == FunctionContext::Host);
+    Builder.setInsertionPointToStart(Module->getBody());
```
Comment (victor-eds) [inline polygeist/tools/cgeist/Lib/clang-mlir.cc:1627]: NIT: Can we have a switch here with a `llvm_unreachable` in the default case instead?

**Review by victor-eds — COMMENTED** [review]
Question: Could we generate `func.func` instead of `llvm.func` instead? What's the point of generating `llvm.func`?

**Comment by whitneywhtsang** [comment]
> Question: Could we generate `func.func` instead of `llvm.func` instead? What's the point of generating `llvm.func`?

I am also unsure why we generate `llvm.func` directly here. Let me give that a try.

**Comment by whitneywhtsang** [comment]
> > Question: Could we generate `func.func` instead of `llvm.func` instead? What's the point of generating `llvm.func`?
> 
> I am also unsure why we generate `llvm.func` directly here. Let me give that a try.

I intended to change `llvm.func` to `func.func`, but it requires non trivial changes, e.g., we cannot create `LLVM::AddressOfOp` of `func::FuncOp`. As there are no known opportunities to change to `func.func`, @victor-eds are you ok if we keep it as is?

**Comment by victor-eds** [comment]
> > > Question: Could we generate `func.func` instead of `llvm.func` instead? What's the point of generating `llvm.func`?
> > 
> > 
> > I am also unsure why we generate `llvm.func` directly here. Let me give that a try.
> 
> I intended to change `llvm.func` to `func.func`, but it requires non trivial changes, e.g., we cannot create `LLVM::AddressOfOp` of `func::FuncOp`. As there are no known opportunities to change to `func.func`, @victor-eds are you ok if we keep it as is?



> > > Question: Could we generate `func.func` instead of `llvm.func` instead? What's the point of generating `llvm.func`?
> > 
> > 
> > I am also unsure why we generate `llvm.func` directly here. Let me give that a try.
> 
> I intended to change `llvm.func` to `func.func`, but it requires non trivial changes, e.g., we cannot create `LLVM::AddressOfOp` of `func::FuncOp`. As there are no known opportunities to change to `func.func`, @victor-eds are you ok if we keep it as is?

Offline discussion: we will keep generating `llvm.func` for the functions named in the set in `GCCall.cc`. We will explore dropping this list in a further PR as we suspect it might not be needed anymore.

**Review by victor-eds — APPROVED** [review]

**Pull request merged**